drivers/adc/adc_cc32xx.c

```diff
@@ -0,0 +1,327 @@
+/*
+ * Copyright (c) 2021 Pavlo Hamov <pasha.gamov@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT ti_cc32xx_adc
+
+#include <errno.h>
+
+#include <drivers/adc.h>
+#include <device.h>
+#include <kernel.h>
+#include <init.h>
+#include <soc.h>
+
+/* Driverlib includes */
+#include <inc/hw_types.h>
+#include <driverlib/pin.h>
+#include <driverlib/rom.h>
+#include <driverlib/rom_map.h>
+#include <driverlib/prcm.h>
+#include <driverlib/adc.h>
+
+#define CHAN_COUNT 4
+
+#define ADC_CONTEXT_USES_KERNEL_TIMER
+#include "adc_context.h"
+
+#define LOG_LEVEL CONFIG_ADC_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(adc_cc32xx);
+
+#define ISR_MASK (ADC_DMA_DONE | ADC_FIFO_OVERFLOW | ADC_FIFO_UNDERFLOW	\
+		  | ADC_FIFO_EMPTY | ADC_FIFO_FULL)
+
+struct adc_cc32xx_data {
+	struct adc_context ctx;
+	const struct device *dev;
+	uint16_t *buffer;
+	uint16_t *repeat_buffer;
+	uint32_t channels;
+	uint8_t offset[CHAN_COUNT];
+	size_t active_channels;
+};
+
+struct adc_cc32xx_cfg {
+	unsigned long base;
+	void (*irq_cfg_func)(void);
+};
+
+static const int s_chPin[CHAN_COUNT] = {
+	PIN_57,
+	PIN_58,
+	PIN_59,
+	PIN_60,
+};
+
+static const int s_channel[CHAN_COUNT] = {
+	ADC_CH_0,
+	ADC_CH_1,
+	ADC_CH_2,
+	ADC_CH_3,
+};
+
+static inline void start_sampling(unsigned long base, int ch)
+{
+	MAP_ADCChannelEnable(base, ch);
+	for (int i = 0; i < 5; i++) {
+		while (!MAP_ADCFIFOLvlGet(base, ch)) {
+		}
+		MAP_ADCFIFORead(base, ch);
+	}
+	MAP_ADCIntClear(base, ch, ISR_MASK);
+	MAP_ADCIntEnable(base, ch, ISR_MASK);
+}
+
+static void adc_context_start_sampling(struct adc_context *ctx)
+{
+	struct adc_cc32xx_data *data =
+		CONTAINER_OF(ctx, struct adc_cc32xx_data, ctx);
+	const struct adc_cc32xx_cfg *config = data->dev->config;
+
+	data->channels = ctx->sequence.channels;
+	data->repeat_buffer = data->buffer;
+
+	for (int i = 0; i < CHAN_COUNT; ++i) {
+		if (ctx->sequence.channels & BIT(i)) {
+			start_sampling(config->base, s_channel[i]);
+		}
+	}
+}
+
+static void adc_context_update_buffer_pointer(struct adc_context *ctx,
+					      bool repeat)
+{
+	struct adc_cc32xx_data *data =
+		CONTAINER_OF(ctx, struct adc_cc32xx_data, ctx);
+
+	if (repeat) {
+		data->buffer = data->repeat_buffer;
+	} else {
+		data->buffer += data->active_channels;
```

Comment:
Since `data->buffer` is now updated only here, it seems that `data->repeat_buffer` became redundant, as you could use just:
```
	if (!repeat) {
		data->buffer += data->active_channels;
	}
```